File: lib/inputs/apt/static.ts
import * as fs from "fs";
import { normalize as normalizePath } from "path";
import { IAptFiles } from "../../analyzer/types";
import { getContentAsString } from "../../extractor";
import { ExtractAction, ExtractedLayers } from "../../extractor/types";
import { streamToString } from "../../stream-utils";

export const getDpkgFileContentAction: ExtractAction = {
  actionName: "dpkg",
  filePathMatches: (filePath) =>
    filePath === normalizePath("/var/lib/dpkg/status"),
  callback: streamToString,
};

export function getDpkgFileContentActionMachine() {
  let dpkgFileContent = "";
  let extFileContent = "";
  try {
    dpkgFileContent = fs.readFileSync("/var/lib/dpkg/status").toString();
  } catch (e) {
    // 
  }
  try {
    extFileContent = fs.readFileSync("/var/lib/apt/extended_states").toString();
  } catch (e) {
    // 
  } 

  return {
    dpkgFile: dpkgFileContent,
    extFile: extFileContent
  };

  // actionName: "dpkg",
  // filePathMatches: (filePath) =>
  //   filePath === normalizePath("/var/lib/dpkg/status"),
  // callback: streamToString,
}

export const getExtFileContentAction: ExtractAction = {
  actionName: "ext",
  filePathMatches: (filePath) =>
    filePath === normalizePath("/var/lib/apt/extended_states"),
  callback: streamToString,
};

export function getAptDbFileContent(
  extractedLayers: ExtractedLayers,
): IAptFiles {
  const dpkgContent = getContentAsString(
    extractedLayers,
    getDpkgFileContentAction,
  );
  const dpkgFile = dpkgContent || "";

  const extContent = getContentAsString(
    extractedLayers,
    getExtFileContentAction,
  );
  const extFile = extContent || "";

  return {
    dpkgFile,
    extFile,
  };
}

File: go/pkg/inputs/apt/static.go
// Package apt provides ExtractActions for dpkg/apt package database files.
package apt

import "github.com/snyk/snyk-docker-plugin/pkg/extractor"

const (
	ActionNameDpkg = "dpkg"
	ActionNameExt  = "dpkg-ext"
)

// DpkgAction extracts /var/lib/dpkg/status.
var DpkgAction = extractor.ExtractAction{
	ActionName:      ActionNameDpkg,
	FilePathMatches: func(p string) bool { return p == "/var/lib/dpkg/status" },
}

// ExtAction extracts /var/lib/apt/extended_states (auto-installed markers).
var ExtAction = extractor.ExtractAction{
	ActionName:      ActionNameExt,
	FilePathMatches: func(p string) bool { return p == "/var/lib/apt/extended_states" },
}

// Actions returns all ExtractActions needed for APT/dpkg analysis.
func Actions() []extractor.ExtractAction { return []extractor.ExtractAction{DpkgAction, ExtAction} }
